File: 0201-0300/0288-Unique Word Abbreviation/0288-Unique Word Abbreviation.cpp
class ValidWordAbbr {
public:
    ValidWordAbbr(vector<string> dictionary) {
        for (string s : dictionary) {
            string key = getKey(s);
            if (table.find(key) != table.end()) {
                if (table[key] != s) {
                    table[key] = "";
                }                
            }
            else {
                table[key] = s;
            }
        }
    }
    
    bool isUnique(string word) {
        string key = getKey(word);
        return table.find(key) == table.end() || table[key] == word;
    }
    
private:
    unordered_map<string, string> table;
    string getKey(string s) {
        if (s.size() <= 2) {
            return s;
        }
        
        return s[0] + to_string(s.size() - 2) + s[s.size() - 1];
    }
};

/**
 * Your ValidWordAbbr object will be instantiated and called as such:
 * ValidWordAbbr obj = new ValidWordAbbr(dictionary);
 * bool param_1 = obj.isUnique(word);
 */
 

File: 0201-0300/0288-Unique Word Abbreviation/0288-Unique Word Abbreviation.py
class ValidWordAbbr:
    
    def __init__(self, dictionary: List[str]):
        self.table = {}
        for word in dictionary:
            key = self.getKey(word)
            if key in self.table:
                if self.table[key] != word:
                    self.table[key] = ''
            else:
                self.table[key] = word
        

    def isUnique(self, word: str) -> bool:
        key = self.getKey(word)
        return key not in self.table or self.table[key] == word
        

    def getKey(self, s):
        if len(s) <= 2:
            return s
        return s[0] + str(len(s) - 2) + s[-1]

# Your ValidWordAbbr object will be instantiated and called as such:
# obj = ValidWordAbbr(dictionary)
# param_1 = obj.isUnique(word)
